src/utils/codeBuilder.ts

```diff
@@ -0,0 +1,309 @@
+        }
+
+        // do substitutions
+        const newFileContents = this.replaceAllTemplateVariables(fileContents, {
+            ...templateVariables,
+            ...quickActionVariables
+        });
+
+        // copy to destination directory
+        const objectApiName =
+            templateVariables[CodeBuilder.TEMPLATE_OBJECT_API_NAME];
+        // file name convention example: Account.view.quickAction-meta.xml
+        const destinationFile = `${objectApiName}.${label.toLocaleLowerCase()}.quickAction-meta.xml`;
+
+        this.writeFileContents(
+            CodeBuilder.QA_DESTINATION_DIR,
+            destinationFile,
+            newFileContents
+        );
+    }
+
+    private copyTemplateFiles(
+        templateVariables: TemplateVariables,
+        template: string,
+        destinationLwc: string
+    ) {
+        CodeBuilder.TEMPLATE_FILE_EXTENSIONS.forEach((extension) => {
+            const templateFilePath = path.join(
+                CodeBuilder.TEMPLATE_DIR,
+                template,
+                `${template}.${extension}`
+            );
+            const fileContents = this.readFileContents(templateFilePath);
+
+            // do substitutions
+            const newFileContents = this.replaceAllTemplateVariables(
+                fileContents,
+                templateVariables
+            );
+
+            // copy to destination directory
+            const destinationDir = path.join(
+                CodeBuilder.LWC_DESTINATION_DIR,
+                destinationLwc
+            );
+            const destinationFile = `${destinationLwc}.${extension}`;
+
+            this.writeFileContents(
+                destinationDir,
+                destinationFile,
+                newFileContents
+            );
+        });
+    }
+
+    private replaceAllTemplateVariables(
+        contents: string,
+        templateVariables: TemplateVariables
+    ) {
+        var newFileContents = contents;
+        for (const key in templateVariables) {
+            if (templateVariables.hasOwnProperty(key)) {
+                const value = templateVariables[key];
+                newFileContents = newFileContents.replace(
+                    `///${key}///`,
+                    value
+                );
+            }
+        }
+        return newFileContents;
+    }
+
+    private readFileContents(filePath: string): string {
+        const extensionFilePath = Uri.joinPath(this.extensionUri, filePath);
+        try {
+            return fs.readFileSync(extensionFilePath.fsPath, 'utf8');
+        } catch (err) {
+            console.log(`Could not read file ${filePath}`, err);
+            return '';
+        }
+    }
+
+    private writeFileContents(
+        dirPath: string,
+        filename: string,
+        content: string
+    ) {
+        // ensure dirPath exists
+        if (!fs.existsSync(dirPath)) {
+            try {
+                fs.mkdirSync(dirPath, { recursive: true });
```

Comment:
I've been a little IDK on this approach. Ideally we're operating from a place where the user has properly bootstrapped their project folder to be synced with the Offline Starter Kit project—where all of these folders should already exist. That it might not exist points to a more fundamental issue to me, than this code not functioning. But then you have to do some sanity checking beforehand, falling through on failure there, etc. ¯\\_(ツ)_/¯ not sure of the right way to handle those boundary checks.

Reply:
I.e. I'm not sure if we shouldn't be making more liberal use of the [`ConfigureProjectCommand.validateProjectFolder()` method](https://github.com/salesforce/salesforcedx-vscode-mobile/blob/7632f668981bd3938045f0d7d3e6be6054dd1980/src/commands/wizard/configureProjectCommand.ts#L247-L310) in the up-front validations of these commands that depend upon a functional project folder. Might be food for thought, for another day.